// @katoid/angular-grid-layout/esm2020/lib/grid.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { KtdGridComponent } from './grid.component';
import { KtdGridItemComponent } from './grid-item/grid-item.component';
import { KtdGridDragHandle } from './directives/drag-handle';
import { KtdGridResizeHandle } from './directives/resize-handle';
import { KtdGridService } from './grid.service';
import { KtdGridItemPlaceholder } from '../public-api';
import * as i0 from "@angular/core";
export class KtdGridModule {
}
KtdGridModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.2.3", ngImport: i0, type: KtdGridModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
KtdGridModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.2.3", ngImport: i0, type: KtdGridModule, declarations: [KtdGridComponent,
        KtdGridItemComponent,
        KtdGridDragHandle,
        KtdGridResizeHandle,
        KtdGridItemPlaceholder], imports: [CommonModule], exports: [KtdGridComponent,
        KtdGridItemComponent,
        KtdGridDragHandle,
        KtdGridResizeHandle,
        KtdGridItemPlaceholder] });
KtdGridModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.2.3", ngImport: i0, type: KtdGridModule, providers: [
        KtdGridService
    ], imports: [CommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.2.3", ngImport: i0, type: KtdGridModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        KtdGridComponent,
                        KtdGridItemComponent,
                        KtdGridDragHandle,
                        KtdGridResizeHandle,
                        KtdGridItemPlaceholder
                    ],
                    exports: [
                        KtdGridComponent,
                        KtdGridItemComponent,
                        KtdGridDragHandle,
                        KtdGridResizeHandle,
                        KtdGridItemPlaceholder
                    ],
                    providers: [
                        KtdGridService
                    ],
                    imports: [
                        CommonModule
                    ]
                }]
        }] });
